### ISSUE TITLE
cycle of death

### ISSUE THREAD

**Issue by piejanssens** [piejanssens/sample-ui5-uaa#3]
@petermuessig do you also experience cds-ui5-plugin reloading over and over again upon running `npm run watch`?

All dependencies are currently up-to-date.

**Comment by petermuessig** [comment]
Hi @piejanssens ,

the `@sap/cds-dk` provides the `cds` binary having a `watch` mode. This watch mode makes use of `node-watch` to determine changes in the sources of the CDS server and restarts the server. This also detects changes in the `app/` folder. In addition, the watch mode also provides a livereload feature for the client-side resources. This is watching for changes inside the `app/` folder and notifies the browser to reload once a resource has been changed. But the livereload is not causing the issue for the cyclic reload. 

The issue is coming from the following case: in your `tsconfig.json` you set the option `composite: true` which leads to the creation of a `tsconfig.tsbuildinfo` file which is used by TypeScript for incremental and faster builds. But this file is detected as a change to the server and `node-watch` detects a change although it is not part of the default extensions or the extensions provided as command line argument.

To solve your issue (temporarily), remove the `composite` option from your `tsconfig.json`! In addition, you can also disable `livereload` and restrict the extensions to monitor:

```sh
cds watch --profile local --livereload false --ext cds
```

This should do the trick to stop continuous reloading of `cds watch`.

Cc: @hschaefer123 @chgeo

**Comment by chgeo** [comment]
Thanks @petermuessig, really good to know.

We get reports of constant restarts every now and then, and this could be the explanation for the TS world.  Does this mean we should exclude `tsconfig.tsbuildinfo` by default?

@daogrady @SAPRogi FYI

**Comment by petermuessig** [comment]
@chgeo - hmm, this would solve the issue for now, but when running real node projects inside the `app` folder there can be more files which trigger the reload of the CDS server. I'm wondering a bit about the ignore regex `/app\/.*\/?webapp\/|node_modules|target|\.cds-services\.json$/` which is used internally and not sure whether this is correct like this. Shouldn't be in any case the content of the app files being ignored? But it also may make sense to restart the CDS server in case of the server configuration of the UI project has been changed. 

What was the main reason to restart the server in case of changes in the UI / `app` folder? The annotations? Couldn't they be included with the `ext` config you can provided initially to the `cds watch`?

**Comment by piejanssens** [comment]
Or... 

-- frontend
---- webapp
---- ui5.yaml
---- package.json
-- backend
---- srv
---- gen
---- .cdsrc.json --> cds-plugin-ui5 config to include app from ../frontend
---- package.json 
-- mta.yaml
-- package.json

**Comment by chgeo** [comment]
Yes, cds files for annotations are the reason for why the app folder is not ignored by default. 

Not sure if you know, but since sap/cds-dk 7.2.0, files `app/**/webapp/**` are ignored by default.  This should address the issues with UI files in default layouts. 

For custom layouts, additional ignore configuration would need to be provided by applications though.

**Comment by piejanssens** [comment]
@petermuessig See https://github.com/piejanssens/sample-ui5-uaa/tree/restructuring
I think for my scenario to work, we need to update cds-plugin-ui5 to make it npm workspace aware?

**Comment by petermuessig** [comment]
> Yes, cds files for annotations are the reason for why the app folder is not ignored by default.
> 
> Not sure if you know, but since sap/cds-dk 7.2.0, files `app/**/webapp/**` are ignored by default. This should address the issues with UI files in default layouts.
> 
> For custom layouts, additional ignore configuration would need to be provided by applications though.

@chgeo would it be possible to create an interface to define this regex somehow? I'm wondering because also some applications may have different structures than `app/myapp/webapp` and this is configured in the `ui5.yaml` and I have the knowledge about this in the `cds-plugin-ui5`. AFAICS, for UI5 applications, only changes to the `package.json` and the `ui5.yaml` might be relevant for a restart of the server. All other files should be ignored. Maybe we can quickly meet next week on this?

**Comment by petermuessig** [comment]
BTW, @chgeo - aren't the extensions are taking into account for those files? If yes, why is the file extension `.tsbuildinfo` recognized at all?

**Comment by petermuessig** [comment]
> @petermuessig See https://github.com/piejanssens/sample-ui5-uaa/tree/restructuring I think for my scenario to work, we need to update cds-plugin-ui5 to make it npm workspace aware?

@piejanssens The resturcturing will most probably solve the issue with the `node-watch` as the sources aren't nested anymore (and personally, I like the mono repo structures more 😄 ). But in general, I made it also running as described above - by removing the `composite` option and restrict the `ext` (optionally) it worked for me properly.

**Comment by chgeo** [comment]
> Maybe we can quickly meet next week on this?

Sure.

> aren't the extensions are taking into account for those files? If yes, why is the file extension `.tsbuildinfo` recognized at all?

Good question.  Need to check.

**Comment by petermuessig** [comment]
@chgeo @daogrady @SAPRogi - I found the root cause for this issue, it is related to the `ignore` regex in the `bin/watch.js` of `@sap/cds-dk` - it needs to escape the path separators which is especially on Windows important as otherwise the backslash is interpret as escaping for the next character.

This was the old regex:

```js
const ignore = RegExp(`app${sep}.*${sep}?webapp${sep}|node_modules|target|\\.cds-services\\.json$`)
```

This is the corrected one:

```js
const ignore = RegExp(`(node_modules\\${sep}|app(\\${sep}.+)?\\${sep}((webapp|dist|target)\\${sep}|\\.cds-services\\.json$|tsconfig\\.json$|.*\\.tsbuildinfo$))`)
```

It ensures to ignore all `node_modules` and the content in the `webapp`, `dist` or `target` folder inside the apps as well as the `.cds-services.json`, `tsconfig.json` and `*.tsbuildinfo` files inside the app folder.

I verified this on my Windows machine and the cycle of death is stopped!

Can you please fix this in the `@sap/cds-dk/bin/watch.js`?

**Comment by petermuessig** [comment]
@chgeo @daogrady @SAPRogi - I suggested a PR to `cap/cds-dk` => `#2188`

**Comment by petermuessig** [comment]
Issue is fixed with `@sap/cds-dk:7.3.0`

Thanks @chgeo @daogrady @SAPRogi
